### PR TITLE
Condense Read and Validation modules in cardano-cli

### PR DESCRIPTION
**Pull request by Jimbo4350**
Two things occur before constructing a tx. You must first read the parts you need and then validate that they can indeed be used in a particular era. We condense these into two module in the cli, with the eventual aim of moving them to cardano-api.

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+import           Data.Word
+
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+--TODO: do this nicely via the API too:
+import qualified Cardano.Binary as CBOR
+import           Data.Text (Text)
+--TODO: following import needed for orphan Eq Script instance
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:77]: This doesn't seem like a TODO.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import qualified Data.List as List
+import qualified Data.Text as Text
+import           Data.Word
+
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+--TODO: do this nicely via the API too:
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:74]: Not sure what this TODO means.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+  | MetadataErrorNotAvailableInEra AnyCardanoEra
+
+renderMetadataError :: MetadataError -> Text
+renderMetadataError (MetadataErrorFile fileErr) =
+  Text.pack $ displayError fileErr
+renderMetadataError (MetadataErrorJsonParseError fp jsonErr) =
+  Text.pack $ "Invalid JSON format in file: " <> show fp <>
+              "\nJSON parse error: " <> jsonErr
+renderMetadataError (MetadataErrorConversionError fp metadataErr) =
+  Text.pack $ "Error reading metadata at: " <> show fp <>
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:101]: Sometimes we say "in file: <fp>" and sometimes we say "at: <fp>".

Is there something different going on?  Or should we use the same language in all cases?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+  | ScriptWitnessErrorExpectedPlutus !FilePath !AnyScriptLanguage
+  | ScriptWitnessErrorReferenceScriptsNotSupportedInEra !AnyCardanoEra
+  | ScriptWitnessErrorScriptData ScriptDataError
+
+renderScriptWitnessError :: ScriptWitnessError -> Text
+renderScriptWitnessError (ScriptWitnessErrorFile err) =
+  Text.pack $ displayError err
+renderScriptWitnessError (ScriptWitnessErrorScriptLanguageNotSupportedInEra (AnyScriptLanguage lang) anyEra) =
+  "The script language " <> Text.pack (show lang) <> " is not supported in the " <>
+  renderEra anyEra <> " era'."
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:172]: Is the apostrophe are `era` intended?

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+                    $ getIsCardanoEraConstraint era' (AnyCardanoEra era')
+    Just _ -> do
+
+      case scriptLanguageSupportedInEra era' anyScriptLanguage of
+        Just sLangInEra ->
+          case languageOfScriptLanguageInEra sLangInEra of
+            SimpleScriptLanguage _v ->
+              -- TODO: We likely need another datatype eg data ReferenceScriptWitness lang
+              -- in order to make this branch unrepresentable.
+              error "readScriptWitness: Should not be possible to specify a simple script"
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:271]: 👍

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+              "\nJSON parse error: " <> jsonErr
+renderScriptDataError (ScriptDataErrorConversion fp sDataJsonErr) =
+  Text.pack $ "Error reading metadata at: " <> show fp <>
+              "\n" <> displayError sDataJsonErr
+renderScriptDataError (ScriptDataErrorValidation fp sDataRangeErr) =
+  Text.pack $ "Error validating script data at: " <> show fp <> ":\n" <>
+              displayError sDataRangeErr
+renderScriptDataError (ScriptDataErrorMetadataDecode fp decoderErr) =
+  Text.pack $ "Error decoding CBOR metadata at: " <> show fp <>
+              " Error: " <> show decoderErr
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:331]: Language "in file" vs "at"

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+
+data CddlError = CddlErrorTextEnv
+                   !(FileError TextEnvelopeError)
+                   !(FileError TextEnvelopeCddlError)
+               | CddlIOError (FileError TextEnvelopeError)
+
+renderCddlError :: CddlError -> Text
+renderCddlError (CddlErrorTextEnv textEnvErr cddlErr) =
+  "Failed to decode neither the cli's serialisation format nor the ledger's \
+  \CDDL serialisation format. TextEnvelope error: " <> Text.pack (displayError textEnvErr) <> "\n" <>
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:483]: Prefer to not use multiline strings.  Prefer using `<>`.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:77]: We actually don't need to import this anymore. I removed it.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:74]: That's actually a comment from Duncan. I think he wanted to expose the CBOR stuff from ledger in a nicer way.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:101]: Nope just an inconsistency. I wouldn't fix that in this PR however. Open an issue if you think its worth fixing.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs (hunk 4 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:172]: Probably not, I'll remove it.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs (hunk 6 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:331]: See comment above

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs (hunk 7 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:483]: Yeah fair enough. We should purge the use of `\`. Can you open an issue for this?

Comment (newhoggy) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Read.hs:481]: https://github.com/input-output-hk/cardano-node/issues/4585

**Review by newhoggy — APPROVED** [review]

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Validate.hs [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+  displayError (ProtocolParametersNotSupported e) =
+    "Transaction protocol parameters are not supported in " <> Text.unpack (renderEra e)
+
+validateProtocolParameters
+  :: CardanoEra era
+  -> Maybe ProtocolParameters
+  -> Either TxProtocolParametersValidationError (BuildTxWith BuildTx (Maybe ProtocolParameters))
+validateProtocolParameters _ Nothing = return (BuildTxWith Nothing)
+validateProtocolParameters era (Just pparams) =
+    case scriptDataSupportedInEra era of
```
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Validate.hs:315]: What's the reason we case match on `scriptDataSupportedInEra` in `validateProtocolParameters`?

**Pull request merged**